File: try_go/person_dao.go
package main

import (
	"database/sql"
)

type person struct {
	ID    int     `json:"id"`
	Name  string  `json:"name"`
	Data string `json:"data"`
}

func (p *person) getPerson(db *sql.DB) error {
	return db.QueryRow("SELECT name, data FROM person WHERE id=$1",
		p.ID).Scan(&p.Name, &p.Data)
}

func (p *person) updatePerson(db *sql.DB) error {
	_, err :=
		db.Exec("UPDATE person SET name=$1, data=$2 WHERE id=$3",
			p.Name, p.Data, p.ID)

	return err
}

func (p *person) deletePerson(db *sql.DB) error {
	_, err := db.Exec("DELETE FROM person WHERE id=$1", p.ID)

	return err
}

func (p *person) createPerson(db *sql.DB) error {
	err := db.QueryRow(
		"INSERT INTO person(name, data) VALUES($1, $2) RETURNING id",
		p.Name, p.Data).Scan(&p.ID)

	if err != nil {
		return err
	}

	return nil
}


func getPersons(db *sql.DB, start, count int) ([]person, error) {
	rows, err := db.Query(
		"SELECT id, name, data FROM person LIMIT $1 OFFSET $2",
		count, start)

	if err != nil {
		return nil, err
	}

	defer rows.Close()

	persons := []person{}

	for rows.Next() {
		var p person
		if err := rows.Scan(&p.ID, &p.Name, &p.Data); err != nil {
			return nil, err
		}
		persons = append(persons, p)
	}

	return persons, nil
}

File: try_rust/src/person_dao.rs
use postgres::Connection;

use std::sync::Mutex;

pub fn insert(db: &Connection, name: &str, data: &str) -> ::postgres::Result<u64> {
    db.execute("INSERT INTO person VALUES (default, $1, $2)", &[&name, &data])
}

pub fn remove(db: &Connection, ids: &[i32]) -> ::postgres::Result<u64> {
    let stmt = db.prepare("DELETE FROM person WHERE id=$1")
        .unwrap();
    for id in ids {
        try!(stmt.execute(&[id]));
    }
    Ok(0)
}

pub fn update(db: &Connection, id: i32, name: &str, data: &str) -> ::postgres::Result<()> {
    let tx: ::postgres::Transaction = db.transaction()
        .unwrap();
    tx.execute("UPDATE person SET name = $1, data = $2 WHERE id = $3", &[&name, &data, &id])
        .unwrap();
    tx.set_commit();
    tx.finish()
}

pub fn show(db: &Connection, arg: Option<&str>) -> ::postgres::Result<Vec<Person>> {
    let s = match arg {
        Some(s) => format!("WHERE name LIKE '%{}%'", s),
        None => "".to_owned(),
    };
    let stmt = db.prepare(&format!("SELECT * FROM person {} ORDER BY id", s))
        .unwrap();
    let rows = stmt.query(&[]).unwrap();
    let size = rows.iter().count();
    let mut results = Vec::with_capacity(size);
    for row in rows {
        let record = Person {
            id: row.get("id"),
            name: row.get("name"),
            data: row.get("data"),
        };
        results.push(record)
    }
    Ok(results)
}

#[derive(RustcEncodable, RustcDecodable)]
pub struct Person {
    id: Option<i32>,
    pub name: String,
    pub data: String,
}

pub fn read(sdb: &Mutex<Connection>, name: Option<&str>) -> Result<Vec<Person>, ()> {
    if let Ok(rs) = show(&*sdb.lock().unwrap(), name) {
        Ok(rs)
    } else {
        Err(())
    }
}

pub fn read_one(sdb: &Mutex<Connection>, id: i32) -> Result<Person, ()> {
    let db = &*sdb.lock().unwrap();
    let stmt = db.prepare("SELECT * FROM person WHERE id = $1")
        .unwrap();
    if let Ok(rows) = stmt.query(&[&id]) {
        let mut iter = rows.iter();
        if iter.len() != 1 {
            return Err(());
        }
        let row = iter.next().unwrap();
        let record = Person {
            id: row.get("id"),
            name: row.get("name"),
            data: row.get("data"),
        };

        Ok(record)
    } else {
        Err(())
    }
}